### ISSUE TITLE
ValueError: Unable to find token seed! Did https://translate.google.com change?

### ISSUE THREAD

**Issue by Mohammed-Shoaib** [pndurette/gTTS#232]
After I send 10-15 requests, I get the following error:

```python
Traceback (most recent call last):
  File "text_to_speech.py", line 25, in <module>
    text_to_speech(data)
  File "text_to_speech.py", line 17, in text_to_speech
    tts.save(f'audio/{key}/def-{i + 1}.mp3')
  File "/home/shoaib/anaconda3/lib/python3.7/site-packages/gtts/tts.py", line 111, in save
    self.write_to_fp(f)
  File "/home/shoaib/anaconda3/lib/python3.7/site-packages/gtts/tts.py", line 124, in write_to_fp
    'tk' : self.token.calculate_token(part)}
  File "/home/shoaib/anaconda3/lib/python3.7/site-packages/gtts_token/gtts_token.py", line 28, in calculate_token
    seed = self._get_token_key()
  File "/home/shoaib/anaconda3/lib/python3.7/site-packages/gtts_token/gtts_token.py", line 59, in _get_token_key
    "Unable to find token seed! Did https://translate.google.com change?"
ValueError: Unable to find token seed! Did https://translate.google.com change?
```

I thought this might be a bug in the latest release, i.e., [2.1.1](https://pypi.org/project/gTTS/2.1.1/) so I tried [1.2.0](https://pypi.org/project/gTTS/1.2.0/) and I still get the same error.

I also thought it could be that my IP address is being blocked. I read up on the [Quotas & limits](https://cloud.google.com/text-to-speech/quotas) and I don't think that should be an issue as well.
* **Characters per request**: 67 (on average), never more than the allowed 5000
* **Requests per minute**: never exceeds 15 since I get the error, allowed 300

#### System information:

* Operating System: Manjaro Linux
* Kernel: Linux 5.8.11-1-MANJARO
* Architecture: x86-64

Thank you for your timely help. :relaxed: :sparkles:

**Comment by thedaynos** [comment]
I started getting this last week periodically. It happened about 10 times to me today.  I have no clue.

**Comment by Mohammed-Shoaib** [comment]
The error is not continuous, it happens randomly from what I can tell. For me it's definitely within 10-15 requests.

Since I observed that it doesn't happen continuously, I just wrote a `while` loop in `python` to try again if the error occurred:

```python
import os
from gtts import gTTS

text = 'Your sentence requiring text to speech'
file_path = 'text.mp3'

while not os.path.exists(file_path) or os.path.getsize(file_path) == 0:
	try:
		tts = gTTS(text=text, lang='en', slow=False)
		tts.save(file_path)
	except Exception as e:
		print(e.message)
```

Obviously, this is a _very_ **bad** solution and I would never recommend it. But it does the job for now as a temporary fix until this issue gets permanently resolved.

**Comment by thedaynos** [comment]
That's interesting. Mine was breaking on the tts.save line, so this wasn't working for me.  This is how I got it working...
`
import os
from gtts import gTTS
count=1
text = 'Your sentence requiring text to speech'
file_path = 'text.mp3'
tts = gTTS(text=text, lang='en', slow=False)
while True:
	try:
		tts.save(file_path)
                break
	except: 
		print('got the issue '+str(count))
                count+=1
`
I have the count in there just out of curiosity, obviously you don't need that.  I just tried this for about 20 minutes and got the error maybe 20% of the time.  Only once did the count get to 2.

Seems to be working without crashing now.

**Comment by afonsosantos** [comment]
Happening with me also, using Python 3.9 on Windows 10.
Used this library before and not issues, check the code here: [https://github.com/afonsosantos/alexis/blob/master/main.py](https://github.com/afonsosantos/alexis/blob/master/main.py)

The error:
```
Traceback (most recent call last):
  File "C:\Users\afons\Desktop\pyvoice\pyVoice\main.py", line 25, in <module>
    respond(voice_data)
  File "C:\Users\afons\Desktop\pyvoice\pyVoice\main.py", line 9, in respond
    date()
  File "C:\Users\afons\Desktop\pyvoice\pyVoice\commands\date_time.py", line 9, in date
    speak('A data de hoje é ' + str(today_date))
  File "C:\Users\afons\Desktop\pyvoice\pyVoice\utils\audio.py", line 51, in speak
    tts.save(audio_file)
  File "C:\Users\afons\Desktop\pyvoice\pyVoice\venv\lib\site-packages\gtts\tts.py", line 295, in save
    self.write_to_fp(f)
  File "C:\Users\afons\Desktop\pyvoice\pyVoice\venv\lib\site-packages\gtts\tts.py", line 251, in write_to_fp
    prepared_requests = self._prepare_requests()
  File "C:\Users\afons\Desktop\pyvoice\pyVoice\venv\lib\site-packages\gtts\tts.py", line 194, in _prepare_requests
    part_tk = self.token.calculate_token(part)
  File "C:\Users\afons\Desktop\pyvoice\pyVoice\venv\lib\site-packages\gtts_token\gtts_token.py", line 28, in calculate_token
    seed = self._get_token_key()
  File "C:\Users\afons\Desktop\pyvoice\pyVoice\venv\lib\site-packages\gtts_token\gtts_token.py", line 58, in _get_token_key
    raise ValueError(
ValueError: Unable to find token seed! Did https://translate.google.com change?
```

**Comment by thedaynos** [comment]
Yes it's crashing at the save.  Take a look at the fix I presented above. It will keep the code working.  Just use a while loop and put the "try" call on the save with a break after.  For the except, you can put anything in there. I just liked to see an error message but it's completely optional.  

This is obviously not ideal but the code is working fine on my end after implementing the while loop.

**Comment by afonsosantos** [comment]
Hi @thedaynos , thanks for the help. I will make that "fix".
Maybe this is a issue related to Python 3.9? What version are you using?

**Comment by thedaynos** [comment]
no problem @afonsosantos ...  I'm using 3.6 on ubuntu.  I updated all of the libraries/packages that I'm using as well.

**Comment by Mohammed-Shoaib** [comment]
@afonsosantos I am using Python `3.8.5`, I doubt it could be an issue related to Python.

@thedaynos You are right, the error isn't that often. For me, the max number of **continuous** failed attempts were 3 and it is occurring on the line `tts.save` function call.

**Comment by badjano** [comment]
found a guy saying to add client=tw-ob in the url ( can´t be sure why, and if it works 100% of the time)
`https://translate.google.com/translate_tts?ie=UTF-8&q=Annoying%20bug&tl=en-US&client=tw-ob`

**Comment by afonsosantos** [comment]
I can make a PR with that change and see if that improves anything.

**Comment by Mohammed-Shoaib** [comment]
I found the [post](https://stackoverflow.com/a/32053452/7442823),
> Add the qualifier '&client=tw-ob' to the end of your query. https://translate.google.com/translate_tts?ie=UTF-8&q=test&tl=zh-TW&client=tw-ob
>
> This answer no longer works consistently. Your ip address will be blocked by google temporarily if you abuse this too much.

Seems like adding `client=tw-ob` doesn't work consistently though.

**Comment by kienpc9x** [comment]
I'm also stuck in this bug.
anyone can fix this

**Comment by afonsosantos** [comment]
@badjano seems like using that link too much gets you locked.

I will for alternative services until this issue is fixed.

**Comment by Fabian42** [comment]
@afonsosantos What is "too much"? Google Translate TTS in general has a limit of 2M characters per day.

**Comment by afonsosantos** [comment]
@Fabian42 as @Mohammed-Shoaib stated:

> Your ip address will be blocked by google temporarily if you abuse this too much.

I don't want the rick of being blocked by using that parameter, and If that does not work either, I will find alternatives.

**Comment by Mohammed-Shoaib** [comment]
@afonsosantos You don't really need to find or use alternatives though.

As mentioned by @thedaynos in his [comment](https://github.com/pndurette/gTTS/issues/232#issuecomment-720693948), you could use the same trick _temporarily_ for now until this gets fixed. I am sure the developers will eventually fix this soon, let's be patient.

**Comment by afonsosantos** [comment]
@Mohammed-Shoaib I will, but I need alternartives in case the issue does not get fixed soon. I don't want to put pressure on the projects devs.
Thanks for all the solutions provided.

**Comment by fidelhuang1** [comment]
i have same problem

**Comment by silvertree-arc** [comment]
I have the same problem, at first was intermittent now all the time. @alexrink96  has posted a good workaround here using the requests library >> (https://github.com/pndurette/gTTS/issues/226#issuecomment-719461765) that uses a different way to access the api and it is working for me at the moment.

**Comment by afonsosantos** [comment]
That looks like a good alternative, even to this library itself. Will test that and report the results.
Thanks @silvertree-arc

**Comment by marvin-w** [comment]
Multiple users in the mycroft chat including me also reported this error. I'll check if I can test the workaround.

**Comment by marvin-w** [comment]
The underlyting gtts-token library was just updated with this change: https://github.com/Boudewijn26/gTTS-token/commit/8d681855214b093c78d5a5da4fa19f92c4e233ef and it was released already.

However, this still doesn't fix the actual problem but instead just hides it in the background. (Retry 5 times or throw error)

**Comment by pndurette** [comment]
Yeah, this is starting to get though to maintain/counter. It's a pretty unorthodox usage of this translate endpoint after all..

But will pull in the updated gtts-token and other fixes and do a release later today.

**Comment by pndurette** [comment]
> found a guy saying to add client=tw-ob in the url ( can´t be sure why, and if it works 100% of the time)
> 
> `https://translate.google.com/translate_tts?ie=UTF-8&q=Annoying%20bug&tl=en-US&client=tw-ob`
> 
> EDIT: BTW, you can use this url to download, so you don´t even need this repo

Over the years, I've had to change the `tw` URL param a few times. Hadn't had to in a while.

**However**, looking at network traffic when using the newer https://translate.google.com, it's pretty obvious that Google has switched to another way of generating speech. It doesn't hit `/translate_tts` anymore. So it probably will disappear eventually. I haven't been able to understand their new method yet.

Considering how many libraries use `gTTS` (or derivatives of it, like Home Assistant, see: https://github.com/home-assistant/core/issues/42911, e.g. https://github.com/home-assistant/core/blob/dev/homeassistant/components/google_translate/tts.py), for sure their SREs are aware of this by now and possibly very annoyed by the usage this generates.

I enjoy a good challenge of trying to find a way to make this work reliability again, but as I was saying, this will become more and more difficult.

**Comment by pndurette** [comment]
`gTTS` has been [updated to `2.1.2`](https://pypi.org/project/gTTS/)!

It updates `gTTS-token` to `1.1.4` and a few other various fixes, see [changelog](https://github.com/pndurette/gTTS/blob/master/CHANGELOG.rst).

I've also reworked the GitHub Action workflows and "marked" with PyTest the ones that try to access the `/translate_tts` URL (most of them) and allowed them to fail for now. Like I stated above, I will _try_ to look into a better way to work around this.

**Comment by Boudewijn26** [comment]
To add to this discussion: I, as maintainer of gTTS-token, also intent to look into this further in order to, if possible, ensure compatibility with the new changes Google is rolling out. This has been complicated by the simple fact I haven't received the new Google Translate yet. It's not at all uncommon for larger corporations to roll out changes gradually and that indeed appears to be happening here. We don't have any insight into which regions get this new "feature" first, so we'll just have to wait for now.

That being said, this is an undocumented API, so we really can't blame Google for changing this without notice. For what I gather Google has turned a blind eye towards this whole thing. It wouldn't be too difficult for them to obfuscate the API to such an amount it'd be tremendously difficult to reverse engineer, which is something they haven't done (yet?) in the many years these projects exist.

[As stated](https://github.com/Boudewijn26/gTTS-token/issues/20#issuecomment-724159634), should this problem persist, everyone is free to comment on https://github.com/Boudewijn26/gTTS-token/issues/20 or raise a new issue there (or continue the discussion here, I suppose).

**Comment by devmaster-terian** [comment]
Ok! We'll be sticked here so as we can hear news about this issue.

**Comment by the-next-ceo** [comment]
`pip install gTTS-token --upgrade`
upgrading the token fixed the issue, at least for me.

**Comment by Boudewijn26** [comment]
Prompted by https://github.com/home-assistant/core/issues/42911#issuecomment-724712969, I've decided to further my investigation. You can follow along https://github.com/Boudewijn26/gTTS-token/pull/23, coincidentally the CI for that PR hit the same issue. Initial findings do point to big changes being needed to maintain compatibility. I'll update the PR regularly, so everyone can follow along.

**Comment by afonsosantos** [comment]
New update from gTTS-token: https://github.com/Boudewijn26/gTTS-token/blob/fix/november-changes/docs/november-2020-translate-changes.md#ladies-and-gentlemen-weve-got-him

**Comment by Boudewijn26** [comment]
Yes, that pull request has now been merged, so the url is https://github.com/Boudewijn26/gTTS-token/blob/master/docs/november-2020-translate-changes.md#ladies-and-gentlemen-weve-got-him. What's perhaps more interesting is what's going on in https://github.com/Boudewijn26/gTTS. Apart from the lookup of locales, I've pretty much gotten it working. I think you can expect a pull request by tomorrow. All in all this wasn't as bad as I was expecting, so I'm pretty relieved.

**Comment by pndurette** [comment]
@Boudewijn26 _Wow_, you're amazing. This is quite a feat of reverse engineering! I wonder if `jQ1olc` is a value that's generated by their build system (during obfuscation, minimization). But it shouldn't be to hard to update (since you made it a constant) if that's the case.

**Comment by pndurette** [comment]
Adding to this (I will put in a more official document or place), I want to:
* Optimize the language retrieval bit since it's not reliable enough anymore. Perhaps switching back to a more static language list (obtained via the same means, but checked in into the repo).
* Strip out all the garbage Python 2 compat. code and make this truly Python 3 with async etc., so things like Home Assistant can use it directly (right now it's some copied bits from <`v2.0.0` of `gTTS`.

**Comment by Boudewijn26** [comment]
Thank you @pndurette, you're too kind. I was wondering the same thing about `jQ1olc`, perhaps it could be a future task for gTTS-token to determine the name of the RPC. 

On the language retrieval: I haven't had a look at the changes there, so it might perhaps be something we can do reliably. I do see a good case for checking them into the repo to prevent flakiness.

**Comment by deltaflyer4747** [comment]
Again not working with gtts 2.1.2 and gtts-token 1.1.4

gets stuck on this:

```
Traceback (most recent call last):
  File "/usr/local/sbin/talk", line 37, in <module>
    tokenizer_cases.period_comma,
  File "/usr/local/lib/python3.5/dist-packages/gtts/tts.py", line 129, in __init__
    langs = tts_langs(self.tld)
  File "/usr/local/lib/python3.5/dist-packages/gtts/lang.py", line 40, in tts_langs
    langs.update(_fetch_langs(tld))
  File "/usr/local/lib/python3.5/dist-packages/gtts/lang.py", line 76, in _fetch_langs
    soup = BeautifulSoup(page.content, 'html.parser')
  File "/usr/local/lib/python3.5/dist-packages/bs4/__init__.py", line 279, in __init__
    markup, from_encoding, exclude_encodings=exclude_encodings)):
  File "/usr/local/lib/python3.5/dist-packages/bs4/builder/_htmlparser.py", line 237, in prepare_markup
    exclude_encodings=exclude_encodings)
  File "/usr/local/lib/python3.5/dist-packages/bs4/dammit.py", line 366, in __init__
    for encoding in self.detector.encodings:
  File "/usr/local/lib/python3.5/dist-packages/bs4/dammit.py", line 264, in encodings
    self.chardet_encoding = chardet_dammit(self.markup)
  File "/usr/local/lib/python3.5/dist-packages/bs4/dammit.py", line 34, in chardet_dammit
    return chardet.detect(s)['encoding']
  File "/usr/local/lib/python3.5/dist-packages/chardet/__init__.py", line 38, in detect
    detector.feed(byte_str)
  File "/usr/local/lib/python3.5/dist-packages/chardet/universaldetector.py", line 211, in feed
    if prober.feed(byte_str) == ProbingState.FOUND_IT:
  File "/usr/local/lib/python3.5/dist-packages/chardet/charsetgroupprober.py", line 71, in feed
    state = prober.feed(byte_str)
  File "/usr/local/lib/python3.5/dist-packages/chardet/sjisprober.py", line 75, in feed
    self.context_analyzer.feed(byte_str[i + 1 - char_len:i + 3
```

**Comment by pndurette** [comment]
Hi all,

I've merged in #244 with @Boudewijn26's new audio download code (thanks again! 🙏)

### Languages
Still working on the languages, which I'm making good progress on. I've identified where the language codes that provide TTS are (vs. all language codes—not all languages in Google Translate provide text-to-speech), I just need to fetch them, with some RegEx magic. Still aiming to push `gTTS` `2.3.0` Nov 14 (GMT-5).

**Comment by marvin-w** [comment]
Thank you for your work on this!

Feel free to tag me when the new release is out so I can open the necessary PRs in the upstream repositories (Home Assistant, Mycroft).

**Comment by pndurette** [comment]
Hi all, @marvin-w & @Boudewijn26 

`gTTS` `2.2.0` has been published! 🎉 
Big changes are #244 & #245

### On languages
> Optimize the language retrieval bit since it's not reliable enough anymore. **Perhaps switching back to a more static language list (obtained via the same means, but checked in into the repo)**.

This is exactly what I did. I figured out a way to programmatically extract it using the new Google Translate (code not yet "GitHub-ready" 😛) and I will add this to the repo, but out of the main code. After that, my idea is to have a GitHub Action run periodically, which could create a PR if it detects a change. A new patch version of `gTTS` could then be released.

### What's next—

* I will rework the inclusion of languages under the hood, so it can be automated (I wanted to get this out ASAP)
* I will rework `gTTSError` and have a less janky error management
* I will work strip out all the old <Python 3.6 stuff, and modernize especially adding coroutine (`async`/`await`) support
---

@marvin-w:
> Feel free to tag me when the new release is out so I can open the necessary PRs in the upstream repositories (Home Assistant, Mycroft).

Here you go! 😄 
As I wrote earlier, I know that HA uses its own version of an old `gTTS`, I'm wondering if `2.2.0` can be integrated as-is, or if it needs to enhanced (i.e. Python 3, async/await, etc). I would love to see HA use `gTTS` directly, and I'm really open to do any changes in this regard.

**Comment by marvin-w** [comment]
Thank you! I'll have a look!

**Comment by marvin-w** [comment]
@pndurette Your lib uses the `requests` package. I think one reason for them to not go with it might be the fact that you aren't using the `aiohttp` package instead. Also, one should probably be able to inject the websession directly into the library.

I'm not quite sure on how I'd tackle the update on HA side, I'll give them a ping in their discord and let them decide. I hope someone of them will reply here :).

**Comment by pndurette** [comment]
`aiohttp` is pretty nice and probably what I should head towards (`gTTS` has been using `requests` since 2014!)

**Comment by hmmbob** [comment]
Seems like the old solution in HA finally broke completely now.... :(

https://github.com/home-assistant/core/issues/43801

**Comment by pndurette** [comment]
@hmmbob Ah yeah that was imminent. Sadly did **not** have any time to work on the above to make gTTS more HA-friendly. 😭

**Comment by hmmbob** [comment]
Just checked with one of the core devs (Frenck): there is no need for you to redo your package to use `aiohttp` instead of `requests`. Work will need to be done in integration the package correctly in the [HA Google TTS component](https://github.com/home-assistant/core/tree/dev/homeassistant/components/google_translate)

Wish I could be of help, but I don't have the programming experience to build this myself....

**Comment by marvin-w** [comment]
@hmmbob Could you properly test it if I do it?

**Comment by hmmbob** [comment]
I'd say yes (it basically involves installing it as custom_component, I guess?)

Have a few people around me that could test too.

**Comment by marvin-w** [comment]
I'd create a feature branch - you'd basically have to check it out and run HA from there and then test google_translate.

**Comment by hmmbob** [comment]
That's ok. Maybe we should switch to [your issue at HA](https://github.com/home-assistant/core/issues/43258) for this effort?

**Comment by devmaster-terian** [comment]
> Seems like the old solution in HA finally broke completely now.... :(
> 
> [home-assistant/core#43801](https://github.com/home-assistant/core/issues/43801)

Yes, now is completely broken... I couldn't use the API since yesterday.

**Comment by justin6478** [comment]
This is the error I get before I add the new gTTS master (that is, with the gtts file and the gtts_token):

Error
An error occurred. Please start Anki while holding down the shift key, which will temporarily disable the add-ons you have installed.
If the issue only occurs when add-ons are enabled, please use the Tools > Add-ons menu item to disable some add-ons and restart Anki, repeating until you discover the add-on that is causing the problem.
When you've discovered the add-on that is causing the problem, please report the issue on the add-on support site.
Debug info:
Anki 2.1.35 (84dcaa86) Python 3.8.0 Qt 5.14.2 PyQt 5.14.2
Platform: Windows 10
Flags: frz=True ao=True sv=1
Add-ons, last update check: 2020-12-02 18:26:16

Caught exception:
Traceback (most recent call last):
  File "aqt\webview.py", line 37, in cmd
  File "aqt\webview.py", line 123, in _onCmd
  File "aqt\webview.py", line 547, in _onBridgeCmd
  File "aqt\editor.py", line 403, in onBridgeCmd
  File "aqt\gui_hooks.py", line 1487, in __call__
  File "anki\hooks.py", line 594, in runFilter
  File "C:\Users\Justin Perez\AppData\Roaming\Anki2\addons21\1128979221\edit.py", line 74, in onFocusLost
    if update_fields(note, field, allFields):
  File "C:\Users\Justin Perez\AppData\Roaming\Anki2\addons21\1128979221\behavior.py", line 270, in update_fields
    fill_sound(hanzi, copy)
  File "C:\Users\Justin Perez\AppData\Roaming\Anki2\addons21\1128979221\behavior.py", line 169, in fill_sound
    s = sound(hanzi, config['speech'])
  File "C:\Users\Justin Perez\AppData\Roaming\Anki2\addons21\1128979221\sound.py", line 51, in sound
    return '[sound:%s]' % AudioDownloader(hanzi, source).download()
  File "C:\Users\Justin Perez\AppData\Roaming\Anki2\addons21\1128979221\tts.py", line 50, in download
    self.func()
  File "C:\Users\Justin Perez\AppData\Roaming\Anki2\addons21\1128979221\tts.py", line 56, in get_google
    tts.save(self.path)
  File "C:\Users\Justin Perez\AppData\Roaming\Anki2\addons21\1128979221\lib\gtts\tts.py", line 243, in save
    self.write_to_fp(f)
  File "C:\Users\Justin Perez\AppData\Roaming\Anki2\addons21\1128979221\lib\gtts\tts.py", line 183, in write_to_fp
    part_tk = self.token.calculate_token(part)
  File "C:\Users\Justin Perez\AppData\Roaming\Anki2\addons21\1128979221\lib\gtts_token\gtts_token.py", line 28, in calculate_token
    seed = self._get_token_key()
  File "C:\Users\Justin Perez\AppData\Roaming\Anki2\addons21\1128979221\lib\gtts_token\gtts_token.py", line 58, in _get_token_key
    raise ValueError(
ValueError: Unable to find token seed! Did https://translate.google.com change?


However, this is what I get when I add the new gTTS master one and delete the other two files: 
Error
An error occurred. Please start Anki while holding down the shift key, which will temporarily disable the add-ons you have installed.
If the issue only occurs when add-ons are enabled, please use the Tools > Add-ons menu item to disable some add-ons and restart Anki, repeating until you discover the add-on that is causing the problem.
When you've discovered the add-on that is causing the problem, please report the issue on the add-on support site.
Debug info:
Anki 2.1.35 (84dcaa86) Python 3.8.0 Qt 5.14.2 PyQt 5.14.2
Platform: Windows 10
Flags: frz=True ao=True sv=1
Add-ons, last update check: 2020-12-02 18:26:16

Caught exception:
Traceback (most recent call last):
  File "aqt\webview.py", line 37, in cmd
  File "aqt\webview.py", line 123, in _onCmd
  File "aqt\webview.py", line 547, in _onBridgeCmd
  File "aqt\editor.py", line 403, in onBridgeCmd
  File "aqt\gui_hooks.py", line 1487, in __call__
  File "anki\hooks.py", line 594, in runFilter
  File "C:\Users\Justin Perez\AppData\Roaming\Anki2\addons21\1128979221\edit.py", line 74, in onFocusLost
    if update_fields(note, field, allFields):
  File "C:\Users\Justin Perez\AppData\Roaming\Anki2\addons21\1128979221\behavior.py", line 270, in update_fields
    fill_sound(hanzi, copy)
  File "C:\Users\Justin Perez\AppData\Roaming\Anki2\addons21\1128979221\behavior.py", line 169, in fill_sound
    s = sound(hanzi, config['speech'])
  File "C:\Users\Justin Perez\AppData\Roaming\Anki2\addons21\1128979221\sound.py", line 51, in sound
    return '[sound:%s]' % AudioDownloader(hanzi, source).download()
  File "C:\Users\Justin Perez\AppData\Roaming\Anki2\addons21\1128979221\tts.py", line 50, in download
    self.func()
  File "C:\Users\Justin Perez\AppData\Roaming\Anki2\addons21\1128979221\tts.py", line 56, in get_google
    tts.save(self.path)
  File "C:\Users\Justin Perez\AppData\Roaming\Anki2\addons21\1128979221\lib\gtts\tts.py", line 243, in save
    self.write_to_fp(f)
  File "C:\Users\Justin Perez\AppData\Roaming\Anki2\addons21\1128979221\lib\gtts\tts.py", line 183, in write_to_fp
    part_tk = self.token.calculate_token(part)
  File "C:\Users\Justin Perez\AppData\Roaming\Anki2\addons21\1128979221\lib\gtts_token\gtts_token.py", line 28, in calculate_token
    seed = self._get_token_key()
  File "C:\Users\Justin Perez\AppData\Roaming\Anki2\addons21\1128979221\lib\gtts_token\gtts_token.py", line 58, in _get_token_key
    raise ValueError(
ValueError: Unable to find token seed! Did https://translate.google.com change?

Please help and explain in simple dumb terms, as I am not a coder in any way, shape or form but a simple man just trying to use anki to learn Chinese. Please help thank you

**Comment by mkanet** [comment]
Please forgive my ignorance.  I am getting this error when using the below card:

```
entity: media_player.living_room_speaker
hold_action:
action: none
show_icon: true
show_name: true
tap_action:
action: call-service
service: tts.google_translate_say
service_data:
    entity_id: media_player.living_room_speaker
    message: My message
type: entity-button
```

However, I am still able to make the same device talk successfully by using below:
```
Call Service:  tts.google_translate_say
entity_id: media_player.living_room_speaker
message: Please remember to close the back door!
```

I'm curious why does one work, but not the other?  They are both making my Google Nest speak.  Will I be able to use the card that's failing now in an upcoming Home Assistant version without changing anything in my configuration?

**Comment by pndurette** [comment]
Just closing this, as it was kept for discussion after the issue was fixed and then bit of an HA direction talk but it's starting to derail a bit.

So cool to see this go to HA, thanks for your work all. I'll sure try to contribute when I can!

**Comment by Hasibulkarimratul** [comment]
Has the Error been fixed? Should I just upgrade gTTS ans gTTS-token?

**Comment by MandarJKulkarni** [comment]
> Has the Error been fixed? Should I just upgrade gTTS ans gTTS-token?

Upgrading the versions worked for me.

**Comment by rajveer01** [comment]
Uninstall and Reinstall of gTTS worked For me.

**Comment by kanishk115** [comment]
> 
> 
> Happening with me also, using Python 3.9 on Windows 10.
> Used this library before and not issues, check the code here: https://github.com/afonsosantos/alexis/blob/master/main.py
> 
> The error:
> 
> ```
> Traceback (most recent call last):
>   File "C:\Users\afons\Desktop\pyvoice\pyVoice\main.py", line 25, in <module>
>     respond(voice_data)
>   File "C:\Users\afons\Desktop\pyvoice\pyVoice\main.py", line 9, in respond
>     date()
>   File "C:\Users\afons\Desktop\pyvoice\pyVoice\commands\date_time.py", line 9, in date
>     speak('A data de hoje é ' + str(today_date))
>   File "C:\Users\afons\Desktop\pyvoice\pyVoice\utils\audio.py", line 51, in speak
>     tts.save(audio_file)
>   File "C:\Users\afons\Desktop\pyvoice\pyVoice\venv\lib\site-packages\gtts\tts.py", line 295, in save
>     self.write_to_fp(f)
>   File "C:\Users\afons\Desktop\pyvoice\pyVoice\venv\lib\site-packages\gtts\tts.py", line 251, in write_to_fp
>     prepared_requests = self._prepare_requests()
>   File "C:\Users\afons\Desktop\pyvoice\pyVoice\venv\lib\site-packages\gtts\tts.py", line 194, in _prepare_requests
>     part_tk = self.token.calculate_token(part)
>   File "C:\Users\afons\Desktop\pyvoice\pyVoice\venv\lib\site-packages\gtts_token\gtts_token.py", line 28, in calculate_token
>     seed = self._get_token_key()
>   File "C:\Users\afons\Desktop\pyvoice\pyVoice\venv\lib\site-packages\gtts_token\gtts_token.py", line 58, in _get_token_key
>     raise ValueError(
> ValueError: Unable to find token seed! Did https://translate.google.com change?
> ```
> 
> **EDIT**: after some analysis, seems to be crashing on the `tts.save` line as below:
> 
> ![image](https://user-images.githubusercontent.com/46262740/97921502-f3d0c000-1d52-11eb-9b24-71cd7e76e003.png)

bro the same is happening with me tooo
I guess there is a server problem from gtts site

**Comment by Shivampurbia** [comment]
tried uninstalling and installing but still the same error.

**Comment by Belgin-Android** [comment]
Uninstall and Reinstalling it Helped Me
